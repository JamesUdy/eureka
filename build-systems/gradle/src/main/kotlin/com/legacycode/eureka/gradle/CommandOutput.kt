package com.legacycode.eureka.gradle

@JvmInline
value class CommandOutput(val content: String)
