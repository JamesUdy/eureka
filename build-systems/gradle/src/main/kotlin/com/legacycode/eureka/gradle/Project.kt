package com.legacycode.eureka.gradle

@JvmInline
value class Project(val name: String)
