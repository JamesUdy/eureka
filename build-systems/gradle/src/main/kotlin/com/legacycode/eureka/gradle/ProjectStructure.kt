package com.legacycode.eureka.gradle

data class ProjectStructure(
  val rootProject: Project,
  val subprojects: List<Project>,
) {
  companion object {
    fun from(output: CommandOutput): ProjectStructure {
      val lines = output.content.lines()
      val rootProject = parseRootProject(lines)

      val subprojects = mutableListOf<Project>()
      for (line in lines) {
        if (isProjectLine(line)) {
          subprojects.add(parseSubproject(line))
        }
      }

      return ProjectStructure(rootProject, subprojects.toList())
    }

    private fun parseRootProject(lines: List<String>): Project {
      val rootProjectLine = lines.firstOrNull { it.startsWith("Root project '") }
      val rootProjectName = rootProjectLine?.substringAfter("Root project '")?.substringBefore("'")!!
      return Project(rootProjectName)
    }

    private fun isProjectLine(line: String): Boolean {
      return line.startsWith("+--- Project") || line.startsWith("\\--- Project")
    }

    private fun parseSubproject(line: String): Project {
      val path = line.substring(line.indexOf(':') + 1).trim()
      return Project(path.substring(0, path.length - 1))
    }
  }
}
