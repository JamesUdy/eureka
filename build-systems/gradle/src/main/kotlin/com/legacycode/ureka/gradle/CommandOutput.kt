package com.legacycode.ureka.gradle

@JvmInline
value class CommandOutput(val content: String)
