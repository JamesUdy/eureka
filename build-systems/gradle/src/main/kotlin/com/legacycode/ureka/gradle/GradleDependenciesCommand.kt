package com.legacycode.ureka.gradle

class GradleDependenciesCommand(
  private val project: Project,
) {
  companion object {
    fun from(project: Project): GradleDependenciesCommand {
      return GradleDependenciesCommand(project)
    }

    fun from(projectStructure: ProjectStructure): List<GradleDependenciesCommand> {
      return projectStructure.subprojects.map(GradleDependenciesCommand::from)
    }
  }

  val text: String
    get() = "./gradlew -q ${project.name}:dependencies"
}
