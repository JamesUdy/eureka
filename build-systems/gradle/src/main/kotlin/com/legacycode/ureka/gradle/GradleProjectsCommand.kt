package com.legacycode.ureka.gradle

import java.io.File

class GradleProjectsCommand(private val projectRoot: File) {
  companion object {
    fun from(projectRoot: File): GradleProjectsCommand {
      return GradleProjectsCommand(projectRoot)
    }
  }

  val text: String
    get() = "${projectRoot.resolve("gradlew")} -q projects"
}
