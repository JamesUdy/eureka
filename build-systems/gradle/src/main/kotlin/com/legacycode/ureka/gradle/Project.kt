package com.legacycode.ureka.gradle

class Project(val name: String) {
    val subprojects: MutableList<Project> = mutableListOf()
    var parentProject: Project? = null
}
