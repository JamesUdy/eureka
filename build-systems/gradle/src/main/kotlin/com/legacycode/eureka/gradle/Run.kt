package com.legacycode.eureka.gradle

import com.legacycode.eureka.gradle.commands.GradleDependenciesCommand
import com.legacycode.eureka.gradle.commands.GradleProjectsCommand
import com.legacycode.eureka.gradle.metrics.Instability
import com.legacycode.eureka.gradle.metrics.Instability.Companion.MAXIMALLY_STABLE
import com.legacycode.eureka.gradle.metrics.Instability.Companion.MAXIMALLY_UNSTABLE
import com.legacycode.eureka.gradle.metrics.Instability.Companion.INDEPENDENT
import java.io.File
import java.util.Locale

private const val RADIX_HEX = 16
private const val CHANNEL_MAX = 255

fun runGradleCommands(projectRoot: File): String {
  val cmdFile = projectRoot.resolve("gradlew")
  val projectsCommand = GradleProjectsCommand(projectRoot)
  val projectsOutput = projectsCommand.execute()

  val projectStructure = ProjectStructure.from(CommandOutput(projectsOutput))
  val gradleDependenciesCommands = GradleDependenciesCommand.from(projectRoot, projectStructure)

  val resolvedDependencies = resolveDependencies(cmdFile, gradleDependenciesCommands)

  return graphvizOutput(projectStructure, resolvedDependencies)
}

private fun resolveDependencies(
  cmdFile: File,
  commands: List<GradleDependenciesCommand>,
): Map<Project, List<SubprojectDependency>> {
  val resolvedDependencies = mutableMapOf<Project, List<SubprojectDependency>>()
  commands.onEach { dependenciesCommand ->
    val project = dependenciesCommand.project
    val output = CommandOutput(dependenciesCommand.execute(cmdFile))
    val dependencies = SubprojectDependency.from(output)
    resolvedDependencies[project] = dependencies.filter { it.name != project.name }
  }
  return resolvedDependencies.toMap()
}

private fun printDebugOutput(resolvedDependencies: Map<Project, List<SubprojectDependency>>) {
  resolvedDependencies.entries.onEach { (project, dependencies) ->
    println("${project.name} (${dependencies.size})")

    dependencies.forEach {
      println("  â†’ ${it.name}")
    }
    println()
  }
}

fun graphvizOutput(
  projectStructure: ProjectStructure,
  subprojectDependencies: Map<Project, List<SubprojectDependency>>,
): String {
  val graphvizBuilder = StringBuilder()

  graphvizBuilder.appendLine("""digraph "${projectStructure.rootProject.name}" {""")
  graphvizBuilder.appendLine("""  node [fontname="Arial"];""")
  graphvizBuilder.appendLine()

  projectStructure.subprojects.onEach { project ->
    graphvizBuilder.appendLine(createNode(project, subprojectDependencies))
  }

  graphvizBuilder.appendLine()
  val projectsWithSubprojectDependencies = subprojectDependencies.filter { it.value.isNotEmpty() }
  projectsWithSubprojectDependencies.onEach { (project, subprojectDependencies) ->
    subprojectDependencies.onEach { subprojectDependency ->
      graphvizBuilder.appendLine(mapProjectToDependency(project, subprojectDependency))
    }
  }

  graphvizBuilder.appendLine("}")

  return graphvizBuilder.toString()
}

private fun createNode(
  project: Project,
  dependencies: Map<Project, List<SubprojectDependency>>,
): String {
  val projectDependencies = dependencies[project]
  val fanOut = projectDependencies?.size ?: 0
  val fanIn = dependencies.flatMap { it.value }.count { it.name == project.name }
  val instability = Instability(fanOut, fanIn)

  val nodeText = """  "${project.name}" [label="${project.name}\n($fanOut, $fanIn, I=${instability.value})""""
  val nodeStyle = nodeStyle(instability)
  return nodeText + nodeStyle
}

private fun nodeStyle(instability: Instability): String {
  val color = when (instability.value) {
    MAXIMALLY_UNSTABLE -> "#01BFFFFF"
    MAXIMALLY_STABLE -> "#FF7F50FF"
    INDEPENDENT -> "#C1CDCDFF"
    else -> {
      val alpha = ((1 - instability.value.toFloat()) * CHANNEL_MAX).toInt()
      "#FF7F50${decimalToHex(alpha)}"
    }
  }

  return """, style="filled", fillcolor="$color"]"""
}

private fun decimalToHex(value: Int): String {
  return value.toString(RADIX_HEX).uppercase(Locale.ENGLISH)
}

private fun mapProjectToDependency(
  project: Project,
  subprojectDependency: SubprojectDependency,
): String {
  return """  "${project.name}" -> "${subprojectDependency.name}""""
}
