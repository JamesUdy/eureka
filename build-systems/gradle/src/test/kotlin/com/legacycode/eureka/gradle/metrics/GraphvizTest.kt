package com.legacycode.eureka.gradle.metrics

import com.legacycode.eureka.gradle.graphvizOutput
import org.approvaltests.Approvals
import org.junit.jupiter.api.Test

class GraphvizTest {
  @Test
  fun `print graphviz dsl`() {
    // given
    val (projectStructure, subprojectDependencies) = eureka

    // when
    val graphviz = graphvizOutput(projectStructure, subprojectDependencies)

    // then
    Approvals.verify(graphviz)
  }
}
