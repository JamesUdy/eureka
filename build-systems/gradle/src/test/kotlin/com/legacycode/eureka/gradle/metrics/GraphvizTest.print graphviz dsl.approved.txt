digraph "tumbleweed" {
  node [fontname="Arial"];

  "android" [label="android\n(3, 2, I=0.60)", style="filled", fillcolor="#FF7F5065"]
  "build-systems-gradle" [label="build-systems-gradle\n(0, 1, I=0)", style="filled", fillcolor="#FF7F50FF"]
  "bytecode-samples" [label="bytecode-samples\n(0, 3, I=0)", style="filled", fillcolor="#FF7F50FF"]
  "bytecode-scanner" [label="bytecode-scanner\n(2, 4, I=0.33)", style="filled", fillcolor="#FF7F50AA"]
  "bytecode-testing" [label="bytecode-testing\n(0, 4, I=0)", style="filled", fillcolor="#FF7F50FF"]
  "cli" [label="cli\n(7, 0, I=1)", style="filled", fillcolor="#01BFFFFF"]
  "filesystem" [label="filesystem\n(1, 2, I=0.33)", style="filled", fillcolor="#FF7F50AA"]
  "vcs" [label="vcs\n(0, 1, I=0)", style="filled", fillcolor="#FF7F50FF"]
  "viz" [label="viz\n(3, 3, I=0.50)", style="filled", fillcolor="#FF7F507F"]
  "web-client-react" [label="web-client-react\n(0, 0, I=âˆž)", style="filled", fillcolor="#C1CDCDFF"]
  "web-server" [label="web-server\n(5, 1, I=0.83)", style="filled", fillcolor="#FF7F502B"]

  "android" -> "bytecode-scanner"
  "android" -> "viz"
  "android" -> "bytecode-testing"
  "bytecode-scanner" -> "bytecode-samples"
  "bytecode-scanner" -> "bytecode-testing"
  "cli" -> "web-server"
  "cli" -> "filesystem"
  "cli" -> "bytecode-scanner"
  "cli" -> "viz"
  "cli" -> "android"
  "cli" -> "build-systems-gradle"
  "cli" -> "bytecode-testing"
  "filesystem" -> "bytecode-samples"
  "viz" -> "bytecode-scanner"
  "viz" -> "bytecode-samples"
  "viz" -> "bytecode-testing"
  "web-server" -> "bytecode-scanner"
  "web-server" -> "filesystem"
  "web-server" -> "vcs"
  "web-server" -> "android"
  "web-server" -> "viz"
}
